### PR TITLE
TINKERPOP3-858 Cleanup after failed plugin install

### PR DESCRIPTION
**Pull request by pluradj**
https://issues.apache.org/jira/browse/TINKERPOP3-858 see manual testing steps in the JIRA

### CODE REVIEW

**Comment by spmallette** [comment]
VOTE: +1

File: gremlin-console/src/main/groovy/org/apache/tinkerpop/gremlin/console/commands/InstallCommand.groovy [hunk 1]
```diff
@@ -53,6 +53,12 @@ class InstallCommand extends CommandSupport {
             final def pluginsThatNeedRestart = grabDeps(dep)
             return "Loaded: " + arguments + (pluginsThatNeedRestart.size() == 0 ? "" : " - restart the console to use $pluginsThatNeedRestart")
         } catch (Exception ex) {
+            if (!(ex instanceof IllegalStateException)) {
+                // IllegalStateException is thrown if a module with the same name is already installed.
+                final def uninstall = new UninstallCommand(shell, mediator)
+                final List<String> module = Collections.singletonList(artifact.getArtifact())
+                uninstall.execute(module)
```
Comment (mhfrantz) [inline gremlin-console/src/main/groovy/org/apache/tinkerpop/gremlin/console/commands/InstallCommand.groovy]: Can uninstall throw an exception?

**Comment by pluradj** [comment]
@dkuppitz `DependencyGrabber.copyDependenciesToPath()` throws two exceptions explicitly, `IOException` and `IllegalStateException`. `RuntimeException` occurs if there is an error grabbing grapes. This is the initial bug report, and its failed directory should be rolled back.

`IOException` occurs if there are problems creating the directories under `ext`. Rollback should occur here, but most likely there isn't even a directory to delete.

`IllegalStateException` occurs if plugin module already exists. For example, `tinkergraph-gremlin` is installed by default out of the box. If the code doesn't catch the `IllegalStateException`, `tinkergraph-gremlin` would get *uninstalled* if the user tried to install it a second time.

File: gremlin-console/src/main/groovy/org/apache/tinkerpop/gremlin/console/commands/InstallCommand.groovy (hunk 1 above)
Comment (pluradj) [inline gremlin-console/src/main/groovy/org/apache/tinkerpop/gremlin/console/commands/InstallCommand.groovy]: No, I didn't see any exceptions declared or thrown in the class.

**Comment by dkuppitz** [comment]
+1

**Comment by okram** [comment]
If @pluradj votes +1, then there will be three. Thus, @pluradj  -- if you think your work is solid, then +1 it and tally the vote and merge. Good luck.

**Comment by pluradj** [comment]
Thanks for the comments. I've centralized the rollback logic into DependencyGrabber, and added a couple unit tests.

**Comment by pluradj** [comment]
Vote: +1

**Comment by okram** [comment]
You have 3 +1 votes. Please merge and if this is going into tp30, don't forget to merge to master as well. Good luck.

**Pull request merged**